Build failure with Ifpack2, Zoltan2, and AMD TPL

At commit 17ccfa78c106fd86812563a4ab3d50ceace9c13a, I configure Trilinos with simply:

```
cmake \
  -DTPL_ENABLE_AMD:BOOL=YES \
  -DBUILD_SHARED_LIBS:BOOL=YES \
  -DTrilinos_ENABLE_Zoltan2:BOOL=YES \
  -DTrilinos_ENABLE_Ifpack2:BOOL=YES \
  ..
```

During build, I get the following error:

```
In file included from /ptmp/Trilinos/packages/zoltan2/src/algorithms/order/Zoltan2_OrderingAlgorithms.hpp:53:0,
                 from /ptmp/Trilinos/packages/zoltan2/src/problems/Zoltan2_OrderingProblem.hpp:54,
                 from /ptmp/Trilinos/packages/ifpack2/src/Ifpack2_AdditiveSchwarz_def.hpp:68,
                 from /ptmp/Trilinos/build/build-zoltan2-amd/packages/ifpack2/src/Ifpack2_AdditiveSchwarz.hpp:2,
                 from /ptmp/Trilinos/packages/ifpack2/src/Ifpack2_Details_Factory_def.hpp:47,
                 from /ptmp/Trilinos/build/build-zoltan2-amd/packages/ifpack2/src/Ifpack2_Details_Factory.hpp:2,
                 from /ptmp/Trilinos/packages/ifpack2/src/Ifpack2_Factory_decl.hpp:48,
                 from /ptmp/Trilinos/build/build-zoltan2-amd/packages/ifpack2/src/Ifpack2_Factory.hpp:1,
                 from /ptmp/Trilinos/packages/ifpack2/src/Ifpack2_Details_LinearSolverFactory_def.hpp:54,
                 from /ptmp/Trilinos/build/build-zoltan2-amd/packages/ifpack2/src/Ifpack2_Details_LinearSolverFactory.hpp:2,
                 from /ptmp/Trilinos/packages/ifpack2/src/Ifpack2_Details_registerLinearSolverFactory.cpp:45:
/ptmp/Trilinos/packages/zoltan2/src/algorithms/order/Zoltan2_AlgAMD.hpp: In instantiation of 'int Zoltan2::AlgAMD<Adapter>::order(const Teuchos::RCP<Zoltan2::OrderingSolution<typename Adapter::lno_t, typename Adapter::gno_t> >&) [with Adapter = Zoltan2::MatrixAdapter<Xpetra::RowMatrix<double, int, long long int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial> >, Xpetra::RowMatrix<double, int, long long int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial> > >; typename Adapter::gno_t = long long int; typename Adapter::lno_t = int]':
/ptmp/Trilinos/packages/zoltan2/src/problems/Zoltan2_OrderingProblem.hpp:229:11:   required from 'void Zoltan2::OrderingProblem<Adapter>::solve(bool) [with Adapter = Zoltan2::XpetraRowMatrixAdapter<Xpetra::RowMatrix<double, int, long long int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial> >, Xpetra::RowMatrix<double, int, long long int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial> > >]'
/ptmp/Trilinos/packages/ifpack2/src/Ifpack2_AdditiveSchwarz_def.hpp:1517:5:   required from 'void Ifpack2::AdditiveSchwarz<MatrixType, LocalInverseType>::setup() [with MatrixType = Tpetra::RowMatrix<double, int, long long int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial> >; LocalInverseType = Ifpack2::Preconditioner<double, int, long long int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial> >]'
/ptmp/Trilinos/packages/ifpack2/src/Ifpack2_AdditiveSchwarz_def.hpp:1134:11:   required from 'void Ifpack2::AdditiveSchwarz<MatrixType, LocalInverseType>::initialize() [with MatrixType = Tpetra::RowMatrix<double, int, long long int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial> >; LocalInverseType = Ifpack2::Preconditioner<double, int, long long int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial> >]'
/ptmp/Trilinos/packages/ifpack2/src/Ifpack2_Details_registerLinearSolverFactory.cpp:76:1:   required from here
/ptmp/Trilinos/packages/zoltan2/src/algorithms/order/Zoltan2_AlgAMD.hpp:154:70: error: no matching function for call to 'Zoltan2::AMDTraits<int>::order(const size_t&, const int*, const long long int*, lno_t*&, double [5], double [20])'
                              edgeIds.getRawPtr(), perm, Control, Info);
                                                                      ^
/ptmp/Trilinos/packages/zoltan2/src/algorithms/order/Zoltan2_AlgAMD.hpp:77:9: note: candidate: int Zoltan2::AMDTraits<int>::order(int, const int*, const int*, int*, double*, double*)
     int order(int n, const int *Ap, const int *Ai, int *perm,
         ^
/ptmp/Trilinos/packages/zoltan2/src/algorithms/order/Zoltan2_AlgAMD.hpp:77:9: note:   no known conversion for argument 3 from 'const long long int*' to 'const int*'
packages/ifpack2/src/CMakeFiles/ifpack2.dir/build.make:62: recipe for target 'packages/ifpack2/src/CMakeFiles/ifpack2.dir/Ifpack2_Details_registerLinearSolverFactory.cpp.o' failed
make[2]: *** [packages/ifpack2/src/CMakeFiles/ifpack2.dir/Ifpack2_Details_registerLinearSolverFactory.cpp.o] Error 1
CMakeFiles/Makefile2:6589: recipe for target 'packages/ifpack2/src/CMakeFiles/ifpack2.dir/all' failed
make[1]: *** [packages/ifpack2/src/CMakeFiles/ifpack2.dir/all] Error 2
Makefile:160: recipe for target 'all' failed
make: *** [all] Error 2
```

> It appears there is a mismatch in types when ETI is off. 

FWIW, configuring with an additional `-DTrilinos_ENABLE_EXPLICIT_INSTANTIATION:BOOL=YES` leads to a similar but slightly different error:
```
In file included from /ptmp/Trilinos/packages/zoltan2/src/algorithms/order/Zoltan2_OrderingAlgorithms.hpp:53:0,
                 from /ptmp/Trilinos/packages/zoltan2/src/problems/Zoltan2_OrderingProblem.hpp:54,
                 from /ptmp/Trilinos/packages/ifpack2/src/Ifpack2_AdditiveSchwarz_def.hpp:68,
                 from /ptmp/Trilinos/build/zoltan2-amd-pthread/packages/ifpack2/src/Ifpack2_AdditiveSchwarz_Serial.cpp:50:
/ptmp/Trilinos/packages/zoltan2/src/algorithms/order/Zoltan2_AlgAMD.hpp: In instantiation of 'int Zoltan2::AlgAMD<Adapter>::order(const Teuchos::RCP<Zoltan2::OrderingSolution<typename Adapter::lno_t, typename Adapter::gno_t> >&) [with Adapter = Zoltan2::MatrixAdapter<Xpetra::RowMatrix<double, int, long long int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial> >, Xpetra::RowMatrix<double, int, long long int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial> > >; typename Adapter::gno_t = long long int; typename Adapter::lno_t = int]':
/ptmp/Trilinos/packages/zoltan2/src/problems/Zoltan2_OrderingProblem.hpp:229:11:   required from 'void Zoltan2::OrderingProblem<Adapter>::solve(bool) [with Adapter = Zoltan2::XpetraRowMatrixAdapter<Xpetra::RowMatrix<double, int, long long int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial> >, Xpetra::RowMatrix<double, int, long long int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial> > >]'
/ptmp/Trilinos/packages/ifpack2/src/Ifpack2_AdditiveSchwarz_def.hpp:1517:5:   required from 'void Ifpack2::AdditiveSchwarz<MatrixType, LocalInverseType>::setup() [with MatrixType = Tpetra::RowMatrix<double, int, long long int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial> >; LocalInverseType = Ifpack2::Preconditioner<double, int, long long int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial> >]'
/ptmp/Trilinos/build/zoltan2-amd-pthread/packages/ifpack2/src/Ifpack2_AdditiveSchwarz_Serial.cpp:61:1:   required from here
/ptmp/Trilinos/packages/zoltan2/src/algorithms/order/Zoltan2_AlgAMD.hpp:154:70: error: no matching function for call to 'Zoltan2::AMDTraits<int>::order(const size_t&, const int*, const long long int*, lno_t*&, double [5], double [20])'
                              edgeIds.getRawPtr(), perm, Control, Info);
                                                                      ^
/ptmp/Trilinos/packages/zoltan2/src/algorithms/order/Zoltan2_AlgAMD.hpp:77:9: note: candidate: int Zoltan2::AMDTraits<int>::order(int, const int*, const int*, int*, double*, double*)
     int order(int n, const int *Ap, const int *Ai, int *perm,
         ^
```

--- a/packages/zoltan2/src/algorithms/order/Zoltan2_AlgAMD.hpp
+++ b/packages/zoltan2/src/algorithms/order/Zoltan2_AlgAMD.hpp
@@ -149,41 +149,6 @@ class AlgAMD : public Algorithm<Adapter>
       // wgts are ignored in AMD
       model->getEdgeList(edgeIds, offsets, wgts);
 
-#if AMD_WITH_LNO_T
-    {
-
-      // SR: We will always call AMD with lno_t
-      AMDTraits<lno_t> AMDobj;
-      double Control[AMD_CONTROL];
-      double Info[AMD_INFO];
-
-      amd_defaults(Control);
-      amd_control(Control);
-
-      lno_t *perm;
-      perm = (lno_t *) (solution->getPermutationRCP().getRawPtr());
-
-      lno_t *amd_edgeIds;
-      // SR: This conversion might throw as we are going from gno_t to
-      // lno_t. Another option is to change local ordering to use gno_t
-      // everywhere and upgrade and use the AMD with gno_t.
-      TPL_Traits<lno_t, const gno_t>::ASSIGN_ARRAY(&amd_edgeIds, edgeIds);
-
-      lno_t amd_nVtx=0;
-      TPL_Traits<lno_t, size_t>::ASSIGN(amd_nVtx, nVtx);
-
-      lno_t result = AMDobj.order(amd_nVtx, offsets.getRawPtr(),
-                             amd_edgeIds, perm, Control, Info);
-
-      if (result != AMD_OK && result != AMD_OK_BUT_JUMBLED)
-          ierr = -1;
-
-      // Delete copies
-      TPL_Traits<lno_t, const gno_t>::DELETE_ARRAY(&amd_edgeIds);
-    }
-#else // AMD_WITH_GNO_T
-    {
-
       // We will always call AMD with SuiteSparse_long
       AMDTraits<SuiteSparse_long> AMDobj;
       double Control[AMD_CONTROL];
@@ -226,8 +191,6 @@ class AlgAMD : public Algorithm<Adapter>
       TPL_Traits<SuiteSparse_long, const lno_t>::DELETE_ARRAY(&amd_offsets);
       TPL_Traits<SuiteSparse_long, const gno_t>::DELETE_ARRAY(&amd_edgeIds);
       delete [] amd_perm;
-     }
-#endif
 
       solution->setHavePerm(true);
       return ierr;

--- a/packages/zoltan2/src/algorithms/order/Zoltan2_AlgAMD.hpp
+++ b/packages/zoltan2/src/algorithms/order/Zoltan2_AlgAMD.hpp
@@ -164,7 +164,7 @@ class AlgAMD : public Algorithm<Adapter>
       SuiteSparse_long *amd_offsets, *amd_edgeIds;
       TPL_Traits<SuiteSparse_long, const lno_t>::ASSIGN_ARRAY(&amd_offsets,
              offsets);
-      // This might through depending on how SuiteSparse was compiled
+      // This might throw depending on how SuiteSparse was compiled
       // with long or long long and the size of both of them
       TPL_Traits<SuiteSparse_long, const gno_t>::ASSIGN_ARRAY(&amd_edgeIds,
              edgeIds);

--- a/packages/zoltan2/src/algorithms/order/Zoltan2_AlgAMD.hpp
+++ b/packages/zoltan2/src/algorithms/order/Zoltan2_AlgAMD.hpp
@@ -53,6 +53,7 @@
 #include <Zoltan2_Algorithm.hpp>
 #include <Zoltan2_GraphModel.hpp>
 #include <Zoltan2_OrderingSolution.hpp>
+#include <Zoltan2_TPLTraits.hpp>
 
 
 #ifdef HAVE_ZOLTAN2_AMD
@@ -83,15 +84,17 @@ class AMDTraits<int>
 };
 
 template <>
-class AMDTraits<long>
+class AMDTraits<SuiteSparse_long>
 {
     public:
-    long order(long n, const long *Ap, const long *Ai, long *perm,
+    long order(SuiteSparse_long n, const SuiteSparse_long *Ap,
+                const SuiteSparse_long *Ai, SuiteSparse_long *perm,
                 double *control, double *info)
     {
         return (amd_l_order(n, Ap, Ai, perm, control, info));
     }
 };
+
 #endif
 
 }
@@ -146,6 +149,10 @@ class AlgAMD : public Algorithm<Adapter>
       // wgts are ignored in AMD
       model->getEdgeList(edgeIds, offsets, wgts);
 
+#if AMD_WITH_LNO_T
+    {
+
+      // SR: We will always call AMD with lno_t
       AMDTraits<lno_t> AMDobj;
       double Control[AMD_CONTROL];
       double Info[AMD_INFO];
@@ -156,11 +163,71 @@ class AlgAMD : public Algorithm<Adapter>
       lno_t *perm;
       perm = (lno_t *) (solution->getPermutationRCP().getRawPtr());
 
-      lno_t result = AMDobj.order(nVtx, offsets.getRawPtr(),
-                             edgeIds.getRawPtr(), perm, Control, Info);
+      lno_t *amd_edgeIds;
+      // SR: This conversion might throw as we are going from gno_t to
+      // lno_t. Another option is to change local ordering to use gno_t
+      // everywhere and upgrade and use the AMD with gno_t.
+      TPL_Traits<lno_t, const gno_t>::ASSIGN_ARRAY(&amd_edgeIds, edgeIds);
+
+      lno_t amd_nVtx=0;
+      TPL_Traits<lno_t, size_t>::ASSIGN(amd_nVtx, nVtx);
+
+      lno_t result = AMDobj.order(amd_nVtx, offsets.getRawPtr(),
+                             amd_edgeIds, perm, Control, Info);
 
       if (result != AMD_OK && result != AMD_OK_BUT_JUMBLED)
           ierr = -1;
+
+      // Delete copies
+      TPL_Traits<lno_t, const gno_t>::DELETE_ARRAY(&amd_edgeIds);
+    }
+#else // AMD_WITH_GNO_T
+    {
+
+      // We will always call AMD with SuiteSparse_long
+      AMDTraits<SuiteSparse_long> AMDobj;
+      double Control[AMD_CONTROL];
+      double Info[AMD_INFO];
+
+      amd_defaults(Control);
+      amd_control(Control);
+
+      // We will use the lno_t for local ordering
+      lno_t *perm;
+      perm = (lno_t *) (solution->getPermutationRCP().getRawPtr());
+
+      SuiteSparse_long *amd_offsets, *amd_edgeIds;
+      TPL_Traits<SuiteSparse_long, const lno_t>::ASSIGN_ARRAY(&amd_offsets,
+             offsets);
+      // This might through depending on how SuiteSparse was compiled
+      // with long or long long and the size of both of them
+      TPL_Traits<SuiteSparse_long, const gno_t>::ASSIGN_ARRAY(&amd_edgeIds,
+             edgeIds);
+
+      SuiteSparse_long amd_nVtx=0;
+      TPL_Traits<SuiteSparse_long, size_t>::ASSIGN(amd_nVtx, nVtx);
+
+      // Allocate a SuiteSparse_long perm
+      SuiteSparse_long *amd_perm = new SuiteSparse_long[amd_nVtx];
+
+      lno_t result = AMDobj.order(amd_nVtx, amd_offsets,
+                             amd_edgeIds, amd_perm, Control, Info);
+
+      if (result != AMD_OK && result != AMD_OK_BUT_JUMBLED)
+          ierr = -1;
+
+      // SR: This conversion might throw as we are going from SuiteSparse_long
+      // to lno_t. Another option is to change local ordering solution
+      // to use gno_t everywhere
+      for (size_t i = 0; i < nVtx; i++)
+          TPL_Traits<lno_t, SuiteSparse_long>::ASSIGN(perm[i], amd_perm[i]);
+
+      // Delete copies
+      TPL_Traits<SuiteSparse_long, const lno_t>::DELETE_ARRAY(&amd_offsets);
+      TPL_Traits<SuiteSparse_long, const gno_t>::DELETE_ARRAY(&amd_edgeIds);
+      delete [] amd_perm;
+     }
+#endif
 
       solution->setHavePerm(true);
       return ierr;

